### PR TITLE
Add gitlab sign in method #343

### PR DESCRIPTION
**Pull request by timurista**
## Add Gitlab SIgnin Method
Added sign in method on main page for gitlab

And adds user from gitlab email:
image
image
image

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [x] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

File: packages/core/src/layout/SignInPage/gitlabProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  const handleLogin = async () => {
+    try {
+      const identity = await gitlabAuthApi.getBackstageIdentity({
+        instantPopup: true,
+      });
+
+      const profile = await gitlabAuthApi.getProfile();
+      onResult({
+        userId: identity?.id || profile.email,
```
Comment (timurista) [inline packages/core/src/layout/SignInPage/gitlabProvider.tsx:35]: GitLab identity is not returning an id in profile Info, so we could use profile email or set the identity id somewhere else.

Comment (Rugvip) [reply packages/core/src/layout/SignInPage/gitlabProvider.tsx:35]: This need to be done in the backend, as this identity needs to be proven by the backstage id token as well. This is for example how it's done for google auth atm: https://github.com/spotify/backstage/blob/2ff4071209a5d0713ee8082578611ae74f6f98d8/plugins/auth-backend/src/providers/google/provider.ts#L145

**Review by Rugvip — COMMENTED** [review]
Thanks! Sorry about the slow reviews during summer 😅

File: packages/core/src/layout/SignInPage/gitlabProvider.tsx (hunk 1 above)
Comment (timurista) [reply packages/core/src/layout/SignInPage/gitlabProvider.tsx:35]: Ok, I will do it there then. Just like that: `const id = profile.email.split('@')[0]; ` should work the same

File: plugins/auth-backend/src/providers/gitlab/provider.ts [hunk 2]
```diff
@@ -78,6 +78,14 @@ export class GitlabAuthProvider implements OAuthProviderHandlers {
       idToken: params.id_token,
     };
 
+    // gitlab provides an id numeric value (123)
+    // as a fallback
+    let id = passportProfile!.id;
+
+    if (profile.email) {
+      id = profile.email.split('@')[0];
```
Comment (timurista) [inline plugins/auth-backend/src/providers/gitlab/provider.ts:86]: @Rugvip using this to set the id, below we return the providerInfo, profile and backstageIdentity with the id

Comment (Rugvip) [reply plugins/auth-backend/src/providers/gitlab/provider.ts:86]: Yep, looks good! It's really the best we can do for now before we have a clear concept of identities.

**Review by Rugvip — APPROVED** [review]
Thanks!

**Comment by timurista** [comment]
great, thank you

File: plugins/auth-backend/src/providers/gitlab/provider.ts (hunk 2 above)
Comment (ldemailly) [reply plugins/auth-backend/src/providers/gitlab/provider.ts:86]: what happens when 2 people have same first part of email (x@domain1 and x@domain2) - this seems risky/should be using fully qualified email if there is no other unique id?

Comment (Rugvip) [reply plugins/auth-backend/src/providers/gitlab/provider.ts:86]: Stuff breaks. This is waiting for something like https://github.com/spotify/backstage/issues/1401, where it's possible to go an look up an actual identity via an external email or id. For now this is just a placeholder to return some kind of meaningful ID, but it doesn't actually do anything yet.

Comment (ldemailly) [reply plugins/auth-backend/src/providers/gitlab/provider.ts:86]: how is removing the domain (without checking it being maybe a base one) helpful though?

**Pull request merged**